File: cortex/autopilot/service.py
"""cortex.autopilot.service — AutopilotService over the Session primitive.

Phase 03 refactor: ``AutopilotService`` is now a thin orchestrator that
wires together:

    * :class:`cortex.session.service.SessionService` for lifecycle ops
      (the SessionRecord is the canonical state),
    * :class:`cortex.autopilot.policies.PolicyEnforcer` for warnings and
      blocks at lifecycle hooks,
    * :mod:`cortex.autopilot.detectors` for the dry-run *preflight*,
    * :mod:`cortex.documenter` when ``finish(auto=True)`` needs to close
      the session via the canonical documenter pipeline.

The service does **not** open sessions — that is the job of
``cortex_create_spec``. ``start`` adopts whatever session is currently
active and surfaces open-time policy warnings.

Public API kept for backwards-compatibility:
    :meth:`start`        — adopt the active session under the configured policy.
    :meth:`preflight`    — dry-run the detector pipeline.
    :meth:`checkpoint`   — append a checkpoint and surface policy warnings.
    :meth:`finish`       — close the active session (auto=True → documenter).
    :meth:`status`       — describe the active or named session.
    :classmethod:`from_project_root` — wire dependencies from a project root.
"""

from __future__ import annotations

import logging
from collections.abc import Callable, Sequence
from pathlib import Path
from typing import Any

from cortex.autopilot.config import load_autopilot_config
from cortex.autopilot.detectors.ambiguous import AmbiguousRequestDetector
from cortex.autopilot.detectors.base import resolve_detectors
from cortex.autopilot.detectors.default import (
    CodeChangeDetector,
    DocsOnlyDetector,
    LargeRefactorDetector,
    NoopDetector,
    QuestionOnlyDetector,
    SecuritySensitiveDetector,
)
from cortex.autopilot.errors import AutopilotError, NoActiveSessionError
from cortex.autopilot.lifecycle import (
    AutopilotCheckpointRequest,
    AutopilotCheckpointResult,
    AutopilotFinishRequest,
    AutopilotFinishResult,
    AutopilotPreflightRequest,
    AutopilotPreflightResult,
    AutopilotStartRequest,
    AutopilotStartResult,
    AutopilotStatusResult,
)
from cortex.autopilot.models import DetectionRequest
from cortex.autopilot.policies import (
    AutopilotMode,
    AutopilotPolicy,
    EnforcementResult,
    EnforcementSeverity,
    PolicyEnforcer,
)
from cortex.session.errors import SessionNotFound
from cortex.session.models import (
    CheckpointSource,
    SessionRecord,
    SessionStatus,
)
from cortex.session.service import SessionService
from cortex.session.storage import SessionStorage
from cortex.workspace.layout import WorkspaceLayout

logger = logging.getLogger(__name__)


# Default detector set; the user can inject a custom list via the constructor.
def _default_detectors() -> list[Any]:
    return [
        AmbiguousRequestDetector(),
        QuestionOnlyDetector(),
        DocsOnlyDetector(),
        SecuritySensitiveDetector(),
        LargeRefactorDetector(),
        CodeChangeDetector(),
        NoopDetector(),
    ]


# Factory signature for the heavyweight AgentMemory dependency. Lazy so
# the simple flows (status, checkpoint, observe-mode finish) never pay
# the AgentMemory startup cost.
MemoryFactory = Callable[[], Any]


class AutopilotService:
    """Apply an :class:`AutopilotPolicy` over a canonical session lifecycle.

    Dependencies are injectable so tests can supply a temporary
    :class:`SessionService` and skip the documenter integration entirely
    by passing ``memory_factory=None`` and calling ``finish(auto=False)``.
    """

    def __init__(
        self,
        session_service: SessionService,
        policy: AutopilotPolicy,
        repo_root: Path,
        *,
        memory_factory: MemoryFactory | None = None,
        detectors: Sequence[Any] | None = None,
    ) -> None:
        self._sessions = session_service
        self._policy = policy
        self._enforcer = PolicyEnforcer(policy)
        self._repo_root = Path(repo_root)
        self._memory_factory = memory_factory
        self._detectors = list(detectors) if detectors is not None else _default_detectors()

    # ── Construction ─────────────────────────────────────────────

    @classmethod
    def from_project_root(
        cls,
        project_root: Path,
        *,
        policy: AutopilotPolicy | None = None,
    ) -> AutopilotService:
        """Wire a service with workspace-discovered defaults.

        - ``SessionService`` is built over the discovered ``WorkspaceLayout``.
        - ``policy`` defaults to ``AutopilotPolicy.from_config(autopilot.yaml)``.
        - ``memory_factory`` lazily instantiates ``AgentMemory`` on first
          ``finish(auto=True)`` call. If memory cannot be built (no
          ``config.yaml``, missing deps), the call surfaces the exception.
        """
        layout = WorkspaceLayout.discover(project_root)
        storage = SessionStorage(layout.sessions_dir)
        session_service = SessionService(storage, layout.repo_root)
        cfg = load_autopilot_config(layout)
        resolved_policy = policy or AutopilotPolicy.from_config(cfg)

        def _memory_factory() -> Any:
            from cortex.core import AgentMemory

            return AgentMemory(config_path=str(layout.config_path))

        return cls(
            session_service=session_service,
            policy=resolved_policy,
            repo_root=layout.repo_root,
            memory_factory=_memory_factory,
        )

    # ── Properties ───────────────────────────────────────────────

    @property
    def policy(self) -> AutopilotPolicy:
        return self._policy

    @property
    def session_service(self) -> SessionService:
        return self._sessions

    # ── start ────────────────────────────────────────────────────

    def start(self, request: AutopilotStartRequest) -> AutopilotStartResult:
        """Adopt the active session under the (optionally overridden) mode.

        Raises :class:`NoActiveSessionError` when no session is active —
        the user must run ``cortex create-spec`` first.
        """
        active = self._require_active()

        if request.mode is not None and request.mode is not self._policy.mode:
            self._policy = _policy_with_mode(self._policy, request.mode)
            self._enforcer = PolicyEnforcer(self._policy)

        results = self._enforcer.on_session_open(active)
        return AutopilotStartResult(
            session=active,
            policy=self._policy,
            warnings=_warnings(results),
        )

    # ── preflight (dry-run of detectors, no mutation) ────────────

    def preflight(self, request: AutopilotPreflightRequest) -> AutopilotPreflightResult:
        """Run the detector pipeline against the request without touching state.

        Useful as a sanity check before the user dives into a heavy task
        ("the spec looks ambiguous — clarify before starting?").
        """
        detection = resolve_detectors(
            self._detectors,
            DetectionRequest(
                user_request=request.user_request,
                changed_files=list(request.changed_files),
                git_diff_stat=request.git_diff_stat,
            ),
        )
        return AutopilotPreflightResult(detection=detection)

    # ── checkpoint ───────────────────────────────────────────────

    def checkpoint(self, request: AutopilotCheckpointRequest) -> AutopilotCheckpointResult:
        """Append a checkpoint to the active session and surface warnings."""
        active = self._require_active()
        try:
            source = CheckpointSource(request.source)
        except ValueError as exc:
            valid = ", ".join(s.value for s in CheckpointSource)
            raise AutopilotError(
                f"unknown checkpoint source {request.source!r}; valid: {valid}"
            ) from exc

        record = self._sessions.checkpoint(
            active.session_id,
            source=source,
            verified_claims=request.verified_claims,
            unverified_claims=request.unverified_claims,
            artifacts_touched=request.artifacts_touched,
            note=request.note,
        )
        new_checkpoint = record.checkpoints[-1]
        results = self._enforcer.on_checkpoint(
            record,
            new_checkpoint,
            files_in_scope=request.files_in_scope,
        )
        return AutopilotCheckpointResult(
            session=record,
            checkpoint=new_checkpoint,
            warnings=_warnings(results),
        )

    # ── finish ───────────────────────────────────────────────────

    def finish(self, request: AutopilotFinishRequest) -> AutopilotFinishResult:
        """Close the session, optionally via the documenter pipeline.

        ``auto=True`` delegates to the canonical
        ``DocumenterPersister.finalize`` (reconstructs context, runs
        verification hooks, writes session note + ADRs, closes).

        ``auto=False`` closes the record without documenting — used by
        the ``observe`` mode when the user plans to run
        ``cortex finish-session`` later.

        Returns a result with ``blocked=True`` and the policy reason when
        :meth:`PolicyEnforcer.on_pre_close` emits a BLOCK.
        """
        session = self._resolve_target_session(request.session_id)
        if session.status is not SessionStatus.OPEN:
            return AutopilotFinishResult(
                session=session,
                documented=False,
                summary=f"session already in status {session.status.value}; no-op",
            )

        pre = self._enforcer.on_pre_close(session)
        blocks = [r for r in pre if r.severity is EnforcementSeverity.BLOCK]
        if blocks:
            return AutopilotFinishResult(
                session=session,
                blocked=True,
                blocked_reason=blocks[0].reason,
                warnings=_warnings(pre),
            )

        if request.auto:
            return self._finish_auto(session, request, warnings=_warnings(pre))
        return self._finish_manual(session, request, warnings=_warnings(pre))

    # ── status ───────────────────────────────────────────────────

    def status(self, session_id: str | None = None) -> AutopilotStatusResult:
        """Describe the session.

        With ``session_id=None`` returns the active session; otherwise
        looks up by id. Returns ``active=False`` when neither resolves.
        """
        if session_id is None:
            session = self._sessions.get_active()
        else:
            try:
                session = self._sessions.get(session_id)
            except SessionNotFound:
                session = None

        if session is None:
            return AutopilotStatusResult(active=False, policy=self._policy)

        inferred = self._sessions.infer_mode(session.checkpoints)
        return AutopilotStatusResult(
            active=True,
            session=session,
            policy=self._policy,
            checkpoint_count=len(session.checkpoints),
            inferred_mode=inferred.value,
        )

    # ── Internals ────────────────────────────────────────────────

    def _require_active(self) -> SessionRecord:
        active = self._sessions.get_active()
        if active is None:
            raise NoActiveSessionError(
                "No active session. Run `cortex create-spec` first to open one."
            )
        return active

    def _resolve_target_session(self, session_id: str | None) -> SessionRecord:
        if session_id is None:
            return self._require_active()
        try:
            return self._sessions.get(session_id)
        except SessionNotFound as exc:
            raise AutopilotError(f"session {session_id!r} not found") from exc

    def _finish_manual(
        self,
        session: SessionRecord,
        request: AutopilotFinishRequest,
        *,
        warnings: list[str],
    ) -> AutopilotFinishResult:
        """Close the record without invoking the documenter."""
        status = _intent_to_status(request.intent)
        updated = self._sessions.close(
            session.session_id,
            status=status,
            documenter_decision=status,
        )
        return AutopilotFinishResult(
            session=updated,
            documented=False,
            summary=f"closed without documenting ({status.value})",
            warnings=warnings,
        )

    def _finish_auto(
        self,
        session: SessionRecord,
        request: AutopilotFinishRequest,
        *,
        warnings: list[str],
    ) -> AutopilotFinishResult:
        """Invoke the canonical documenter to close + persist + index."""
        if self._memory_factory is None:
            raise AutopilotError(
                "finish(auto=True) requires a memory_factory to invoke the documenter; "
                "either inject one or use finish(auto=False)."
            )
        from cortex.documenter import (
            DocumenterPersister,
            FinishOverrides,
            ReconstructionInput,
            Reconstructor,
        )
        from cortex.session.verification import VerificationRunner

        memory = self._memory_factory()
        reconstructor = Reconstructor(
            session_service=self._sessions,
            verification_runner=VerificationRunner(repo_root=self._repo_root),
            repo_root=self._repo_root,
        )
        out = reconstructor.reconstruct(ReconstructionInput(session_id=session.session_id))

        forced = _intent_to_forced_status(request.intent)
        persister = DocumenterPersister(
            note_service=memory._note_service,
            session_service=self._sessions,
            vault_path=memory._vault_path_resolved,
        )
        result = persister.finalize(
            out,
            overrides=FinishOverrides(forced_status=forced),
        )

        # Reload the session to capture the now-closed state.
        refreshed = self._sessions.get(session.session_id)
        return AutopilotFinishResult(
            session=refreshed,
            documented=not result.already_closed,
            session_note_path=str(result.session_note_path) if result.session_note_path else None,
            adrs_created=[str(p) for p in result.adrs_created],
            summary=result.summary,
            warnings=warnings,
        )


# ── Module helpers ───────────────────────────────────────────────


def _warnings(results: list[EnforcementResult]) -> list[str]:
    return [r.reason for r in results if r.severity is EnforcementSeverity.WARN]


def _policy_with_mode(policy: AutopilotPolicy, mode: AutopilotMode) -> AutopilotPolicy:
    """Return a copy of ``policy`` with the mode replaced.

    Higher modes flip the warning flags on; lowering to OBSERVE turns
    them off so the new mode is fully consistent.
    """
    is_observe = mode is AutopilotMode.OBSERVE
    return AutopilotPolicy(
        mode=mode,
        budget_profile=policy.budget_profile,
        pre_commit_verification=mode is AutopilotMode.AUTOPILOT,
        out_of_scope_warning=not is_observe,
        warn_on_security_summary=not is_observe,
        auto_checkpoint_threshold_files=policy.auto_checkpoint_threshold_files,
        auto_checkpoint_threshold_minutes=policy.auto_checkpoint_threshold_minutes,
    )


def _intent_to_status(intent: str) -> SessionStatus:
    """Translate a finish ``intent`` to a :class:`SessionStatus`."""
    normalized = intent.strip().lower()
    if normalized in {"handoff"}:
        return SessionStatus.HANDOFF
    if normalized in {"abandoned", "abandon"}:
        return SessionStatus.ABANDONED
    return SessionStatus.CLOSED


def _intent_to_forced_status(intent: str) -> SessionStatus | None:
    """Translate a finish ``intent`` to a forced status override.

    ``"closed"`` means *no override* — let the documenter pick CLOSED or
    HANDOFF based on the reconstruction.
    """
    normalized = intent.strip().lower()
    if normalized in {"handoff"}:
        return SessionStatus.HANDOFF
    if normalized in {"abandoned", "abandon"}:
        return SessionStatus.ABANDONED
    return None


__all__ = ["AutopilotService"]

File: rust/crates/cortex-autopilot/src/service.rs
//! Puerto de `cortex.autopilot.service` — `AutopilotService` (Cierre T3).
//!
//! Orquestador delgado que cablea:
//! - [`cortex_app::session::service::SessionService`] NATIVO para el ciclo
//!   de vida (el `SessionRecord` es el estado canónico),
//! - la capa de decisión P12B-5 ([`crate::policies`] + [`crate::detectors`])
//!   para warnings/bloqueos en los hooks y el dry-run *preflight*,
//! - un backend opcional [`DocumenterFinalize`] para `finish(auto=True)`
//!   (equivalente del `memory_factory` perezoso del oráculo; sin él el
//!   cierre automático devuelve fallo EXPLÍCITO con el mensaje exacto).
//!
//! El servicio NO abre sesiones: `start` ADOPTA la sesión activa
//! (`cortex create-spec` es quien la crea).

use std::path::{Path, PathBuf};
use std::sync::Arc;

use cortex_app::session::service::SessionService;
use cortex_app::session::{
    infer_mode, Checkpoint as NativeCheckpoint, CheckpointSource, SessionRecord, SessionStatus,
    SessionStorage,
};

use crate::detectors::{default_detectors, resolve_detectors, AutopilotDetector};
use crate::models::DetectionRequest;
use crate::policies::{AutopilotMode, AutopilotPolicy, EnforcementSeverity, PolicyEnforcer};
use crate::session_models as decision;

// ── Errores ──────────────────────────────────────────────────────────────────

/// Excepciones del módulo autopilot relevantes para la paridad de mensajes:
/// `NoActiveSessionError`, `AutopilotError` genérico y
/// `cortex.session.errors.SessionNotFound`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ServiceError {
    NoActiveSession(String),
    Autopilot(String),
    SessionNotFound(String),
}

impl std::fmt::Display for ServiceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NoActiveSession(m) | Self::Autopilot(m) | Self::SessionNotFound(m) => {
                f.write_str(m)
            }
        }
    }
}
impl std::error::Error for ServiceError {}

const NO_ACTIVE_SESSION_MSG: &str =
    "No active session. Run `cortex create-spec` first to open one.";

// ── Backend documenter (finish auto) ────────────────────────────────────────

/// Resultado de `DocumenterPersister.finalize(out, overrides)`.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct FinalizeOutcome {
    pub already_closed: bool,
    pub session_note_path: Option<String>,
    pub adrs_created: Vec<String>,
    pub summary: String,
}

/// Puerto del pipeline documenter canónico (`Reconstructor` +
/// `DocumenterPersister.finalize`). Producción: cortex-app::documenter +
/// NoteService; gates: stubs deterministas.
pub trait DocumenterFinalize: Send {
    fn finalize(
        &mut self,
        session_id: &str,
        forced_status: Option<&str>,
    ) -> Result<FinalizeOutcome, String>;
}

// ── Outcomes ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct StartOutcome {
    pub session: SessionRecord,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct PreflightOutcome {
    pub detection: crate::models::DetectionResult,
}

#[derive(Debug, Clone)]
pub struct CheckpointOutcome {
    pub session: SessionRecord,
    pub checkpoint: NativeCheckpoint,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct FinishOutcome {
    pub session: SessionRecord,
    pub documented: bool,
    pub blocked: bool,
    pub blocked_reason: String,
    pub session_note_path: Option<String>,
    pub adrs_created: Vec<String>,
    pub summary: String,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct StatusOutcome {
    pub active: bool,
    pub session: Option<SessionRecord>,
    pub checkpoint_count: usize,
    /// `SessionMode.value` inferido (None cuando no hay sesión activa).
    pub inferred_mode: Option<String>,
}

// ── Servicio ─────────────────────────────────────────────────────────────────

pub struct AutopilotService {
    sessions: SessionService,
    policy: AutopilotPolicy,
    enforcer: PolicyEnforcer,
    repo_root: PathBuf,
    finisher: Option<Box<dyn DocumenterFinalize + Send>>,
    detectors: Vec<Box<dyn AutopilotDetector + Send>>,
}

impl AutopilotService {
    /// Cableo completo inyectable (espejo de `__init__`; `detectors` usa
    /// siempre el set default — ningún gate inyecta otros).
    pub fn new(
        sessions: SessionService,
        policy: AutopilotPolicy,
        repo_root: &Path,
        finisher: Option<Box<dyn DocumenterFinalize + Send>>,
    ) -> Self {
        Self {
            sessions,
            enforcer: PolicyEnforcer::new(policy.clone()),
            policy,
            repo_root: repo_root.to_path_buf(),
            finisher,
            detectors: default_detectors(),
        }
    }

    /// `from_project_root`: layout descubierto + storage nativo +
    /// policy desde `autopilot.yaml` (+ backend documenter inyectable).
    pub fn from_project_root(
        project_root: &Path,
        finisher: Option<Box<dyn DocumenterFinalize + Send>>,
    ) -> Result<Self, crate::config::ConfigError> {
        let layout = cortex_workspace::WorkspaceLayout::discover(project_root);
        let storage = SessionStorage::new(layout.repo_root.join(".cortex").join("sessions"));
        let session_service = SessionService::new(storage, &layout.repo_root);
        let cfg = crate::config::load_autopilot_config(&layout)?;
        let clock: Arc<dyn cortex_enterprise::clock::Clock> =
            Arc::new(cortex_enterprise::clock::SystemClock);
        let resolved_policy = AutopilotPolicy::from_config(&cfg, clock)
            .map_err(|e| crate::config::ConfigError(e.to_string()))?;
        Ok(Self::new(
            session_service,
            resolved_policy,
            &layout.repo_root,
            finisher,
        ))
    }

    // ── Properties ───────────────────────────────────────────────

    pub fn policy(&self) -> &AutopilotPolicy {
        &self.policy
    }

    pub fn session_service(&self) -> &SessionService {
        &self.sessions
    }

    pub fn repo_root(&self) -> &Path {
        &self.repo_root
    }

    // ── start ────────────────────────────────────────────────────

    /// Adoptar la sesión activa bajo el modo (opcionalmente sobrescrito).
    ///
    /// `Err(ServiceError::NoActiveSession)` si no hay sesión activa.
    pub fn start(&mut self, mode: Option<AutopilotMode>) -> Result<StartOutcome, ServiceError> {
        let active = self.require_active()?;

        if let Some(m) = mode {
            if m != self.policy.mode {
                self.policy = policy_with_mode(&self.policy, m);
                self.enforcer = PolicyEnforcer::new(self.policy.clone());
            }
        }

        let dec_active = to_decision_record(&active);
        let results = self.enforcer.on_session_open(&dec_active, None);
        Ok(StartOutcome {
            session: active,
            warnings: warnings_of(&results),
        })
    }

    // ── preflight (dry-run de detectors, sin mutación) ───────────

    pub fn preflight(
        &self,
        user_request: Option<&str>,
        changed_files: &[String],
        git_diff_stat: Option<&str>,
    ) -> PreflightOutcome {
        let detection = resolve_detectors(
            &self.detectors,
            &DetectionRequest {
                user_request: user_request.map(str::to_string),
                changed_files: changed_files.to_vec(),
                git_diff_stat: git_diff_stat.map(str::to_string),
                session_state: None,
            },
        );
        PreflightOutcome { detection }
    }

    // ── checkpoint ───────────────────────────────────────────────

    #[allow(clippy::too_many_arguments)]
    pub fn checkpoint(
        &mut self,
        source: &str,
        verified_claims: Vec<String>,
        unverified_claims: Vec<String>,
        artifacts_touched: Vec<String>,
        note: &str,
        files_in_scope: Option<Vec<String>>,
    ) -> Result<CheckpointOutcome, ServiceError> {
        let active = self.require_active()?;
        let parsed_source = parse_checkpoint_source(source).ok_or_else(|| {
            ServiceError::Autopilot(format!(
                "unknown checkpoint source {}; valid: {}",
                py_repr(source),
                VALID_SOURCES.join(", ")
            ))
        })?;

        let record = self
            .sessions
            .checkpoint(
                &active.session_id,
                parsed_source,
                verified_claims,
                unverified_claims,
                artifacts_touched,
                note,
                None, // autopilot no emite fase COMPOSED
            )
            .map_err(ServiceError::Autopilot)?;
        let new_checkpoint = record
            .checkpoints
            .last()
            .cloned()
            .expect("checkpoint recién agregado");

        let dec_record = to_decision_record(&record);
        let dec_checkpoint = dec_record
            .checkpoints
            .last()
            .cloned()
            .expect("checkpoint espejado");
        let results =
            self.enforcer
                .on_checkpoint(&dec_record, &dec_checkpoint, files_in_scope.as_deref());
        Ok(CheckpointOutcome {
            session: record,
            checkpoint: new_checkpoint,
            warnings: warnings_of(&results),
        })
    }

    // ── finish ───────────────────────────────────────────────────

    pub fn finish(
        &mut self,
        session_id: Option<&str>,
        auto: bool,
        intent: &str,
    ) -> Result<FinishOutcome, ServiceError> {
        let session = self.resolve_target_session(session_id)?;
        if session.status != SessionStatus::Open {
            let summary = format!(
                "session already in status {}; no-op",
                session.status.as_str()
            );
            return Ok(FinishOutcome {
                session,
                documented: false,
                blocked: false,
                blocked_reason: String::new(),
                session_note_path: None,
                adrs_created: vec![],
                summary,
                warnings: vec![],
            });
        }

        let pre = self.enforcer.on_pre_close(&to_decision_record(&session));
        let blocks: Vec<&crate::policies::EnforcementResult> = pre
            .iter()
            .filter(|r| r.severity == EnforcementSeverity::Block)
            .collect();
        if !blocks.is_empty() {
            let warnings = warnings_of(&pre);
            return Ok(FinishOutcome {
                session,
                documented: false,
                blocked: true,
                blocked_reason: blocks[0].reason.clone(),
                session_note_path: None,
                adrs_created: vec![],
                summary: String::new(),
                warnings,
            });
        }

        let warnings = warnings_of(&pre);
        if auto {
            self.finish_auto(&session, intent, warnings)
        } else {
            self.finish_manual(&session, intent, warnings)
        }
    }

    // ── status ───────────────────────────────────────────────────

    pub fn status(&mut self, session_id: Option<&str>) -> Result<StatusOutcome, ServiceError> {
        let found = match session_id {
            None => self.sessions.get_active(),
            Some(sid) => self.sessions.get(sid).ok(),
        };

        let Some(session) = found else {
            return Ok(StatusOutcome {
                active: false,
                session: None,
                checkpoint_count: 0,
                inferred_mode: None,
            });
        };

        let inferred = infer_mode(&session.checkpoints);
        Ok(StatusOutcome {
            active: true,
            checkpoint_count: session.checkpoints.len(),
            inferred_mode: Some(mode_value(inferred).to_string()),
            session: Some(session),
        })
    }

    // ── Internos ─────────────────────────────────────────────────

    fn require_active(&self) -> Result<SessionRecord, ServiceError> {
        self.sessions
            .get_active()
            .ok_or_else(|| ServiceError::NoActiveSession(NO_ACTIVE_SESSION_MSG.to_string()))
    }

    fn resolve_target_session(
        &self,
        session_id: Option<&str>,
    ) -> Result<SessionRecord, ServiceError> {
        match session_id {
            None => self.require_active(),
            Some(sid) => self.sessions.get(sid).map_err(|_| {
                ServiceError::Autopilot(format!("session {} not found", py_repr(sid)))
            }),
        }
    }

    /// Cerrar SIN invocar el documenter.
    fn finish_manual(
        &mut self,
        session: &SessionRecord,
        intent: &str,
        warnings: Vec<String>,
    ) -> Result<FinishOutcome, ServiceError> {
        let status = intent_to_status(intent);
        let updated = self
            .sessions
            .close(&session.session_id, status, status, None, vec![])
            .map_err(ServiceError::Autopilot)?;
        Ok(FinishOutcome {
            summary: format!("closed without documenting ({})", status.as_str()),
            session: updated,
            documented: false,
            blocked: false,
            blocked_reason: String::new(),
            session_note_path: None,
            adrs_created: vec![],
            warnings,
        })
    }

    /// Cierre vía pipeline documenter canónico.
    fn finish_auto(
        &mut self,
        session: &SessionRecord,
        intent: &str,
        warnings: Vec<String>,
    ) -> Result<FinishOutcome, ServiceError> {
        let forced = intent_to_forced_status(intent);
        let outcome = match self.finisher.as_mut() {
            None => {
                return Err(ServiceError::Autopilot(
                    "finish(auto=True) requires a memory_factory to invoke the documenter; \
                     either inject one or use finish(auto=False)."
                        .to_string(),
                ));
            }
            Some(f) => f
                .finalize(&session.session_id, forced)
                .map_err(ServiceError::Autopilot)?,
        };

        // Recargar la sesión para capturar el estado ya cerrado.
        let refreshed = self
            .sessions
            .get(&session.session_id)
            .map_err(ServiceError::Autopilot)?;
        Ok(FinishOutcome {
            session: refreshed,
            documented: !outcome.already_closed,
            session_note_path: outcome.session_note_path,
            adrs_created: outcome.adrs_created,
            summary: outcome.summary,
            blocked: false,
            blocked_reason: String::new(),
            warnings,
        })
    }
}

// ── Helpers de módulo ────────────────────────────────────────────────────────

/// Valores de `CheckpointSource` en el orden del enum Python (mensaje
/// "valid:" del error contractual).
pub const VALID_SOURCES: &[&str] = &[
    "cortex-sync",
    "cortex-SDDwork",
    "cortex-code-explorer",
    "cortex-code-implementer",
    "cortex-code-designer",
    "user-skill",
    "ide-hook",
    "manual",
    "ci-bot",
];

fn parse_checkpoint_source(raw: &str) -> Option<CheckpointSource> {
    Some(match raw {
        "cortex-sync" => CheckpointSource::CortexSync,
        "cortex-SDDwork" => CheckpointSource::CortexSddwork,
        "cortex-code-explorer" => CheckpointSource::CortexCodeExplorer,
        "cortex-code-implementer" => CheckpointSource::CortexCodeImplementer,
        "cortex-code-designer" => CheckpointSource::CortexCodeDesigner,
        "user-skill" => CheckpointSource::UserSkill,
        "ide-hook" => CheckpointSource::IdeHook,
        "manual" => CheckpointSource::Manual,
        "ci-bot" => CheckpointSource::CiBot,
        _ => return None,
    })
}

fn warnings_of(results: &[crate::policies::EnforcementResult]) -> Vec<String> {
    results
        .iter()
        .filter(|r| r.severity == EnforcementSeverity::Warn)
        .map(|r| r.reason.clone())
        .collect()
}

/// Copia de la policy con el modo reemplazado (flags consistentes con el
/// modo nuevo: OBSERVE apaga warnings; AUTOPILOT prende pre-commit).
fn policy_with_mode(policy: &AutopilotPolicy, mode: AutopilotMode) -> AutopilotPolicy {
    let is_observe = mode == AutopilotMode::Observe;
    let mut p = AutopilotPolicy::new(
        mode,
        policy.budget_profile.clone(),
        mode == AutopilotMode::Autopilot,
        !is_observe,
        policy.auto_checkpoint_threshold_files,
        policy.auto_checkpoint_threshold_minutes,
        Arc::new(cortex_enterprise::clock::SystemClock),
    )
    .expect("policy ya validada");
    p.warn_on_security_summary = !is_observe;
    p
}

/// `intent` → `SessionStatus` terminal.
fn intent_to_status(intent: &str) -> SessionStatus {
    let normalized = intent.trim().to_lowercase();
    match normalized.as_str() {
        "handoff" => SessionStatus::Handoff,
        "abandoned" | "abandon" => SessionStatus::Abandoned,
        _ => SessionStatus::Closed,
    }
}

/// `intent` → override de status para el documenter (`None` = decidir solo).
fn intent_to_forced_status(intent: &str) -> Option<&'static str> {
    let normalized = intent.trim().to_lowercase();
    match normalized.as_str() {
        "handoff" => Some("handoff"),
        "abandoned" | "abandon" => Some("abandoned"),
        _ => None,
    }
}

/// `str(SessionMode)` del oráculo ("ci-review" incluido).
fn mode_value(mode: cortex_app::session::SessionMode) -> &'static str {
    match mode {
        cortex_app::session::SessionMode::Unknown => "unknown",
        cortex_app::session::SessionMode::Managed => "managed",
        cortex_app::session::SessionMode::Observed => "observed",
        cortex_app::session::SessionMode::Byo => "byo",
        cortex_app::session::SessionMode::Composed => "composed",
        cortex_app::session::SessionMode::CiReview => "ci-review",
    }
}

/// `repr(str)` de Python (comillas simples por defecto).
pub(crate) fn py_repr(s: &str) -> String {
    if s.contains('\'') && !s.contains('"') {
        format!("\"{}\"", s.replace('\\', "\\\\").replace('"', "\\\""))
    } else {
        format!("'{}'", s.replace('\\', "\\\\").replace('\'', "\\'"))
    }
}

// ── Puente native → decisión (P12B-5 consume sus propios tipos) ─────────────

fn parse_ts(iso: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::DateTime::parse_from_rfc3339(iso)
        .map(|dt| dt.with_timezone(&chrono::Utc))
        .unwrap_or_else(|_| chrono::Utc::now())
}

fn to_decision_status(s: SessionStatus) -> decision::SessionStatus {
    match s {
        SessionStatus::Open => decision::SessionStatus::Open,
        SessionStatus::Closed => decision::SessionStatus::Closed,
        SessionStatus::Handoff => decision::SessionStatus::Handoff,
        SessionStatus::Abandoned => decision::SessionStatus::Abandoned,
    }
}

fn to_decision_checkpoint(cp: &NativeCheckpoint) -> decision::Checkpoint {
    decision::Checkpoint {
        timestamp: parse_ts(&cp.timestamp),
        source: match cp.source {
            CheckpointSource::CortexSync => decision::CheckpointSource::CortexSync,
            CheckpointSource::CortexSddwork => decision::CheckpointSource::CortexSddwork,
            CheckpointSource::CortexCodeExplorer => decision::CheckpointSource::CortexCodeExplorer,
            CheckpointSource::CortexCodeImplementer => {
                decision::CheckpointSource::CortexCodeImplementer
            }
            CheckpointSource::CortexCodeDesigner => decision::CheckpointSource::CortexCodeDesigner,
            CheckpointSource::UserSkill => decision::CheckpointSource::UserSkill,
            CheckpointSource::IdeHook => decision::CheckpointSource::IdeHook,
            CheckpointSource::Manual => decision::CheckpointSource::Manual,
            CheckpointSource::CiBot => decision::CheckpointSource::CiBot,
        },
        verified_claims: cp.verified_claims.clone(),
        unverified_claims: cp.unverified_claims.clone(),
        artifacts_touched: cp.artifacts_touched.clone(),
        note: cp.note.clone(),
    }
}

/// Espejo fiel del record nativo en los tipos mínimos de la capa de
/// decisión (solo campos consumidos por policies/lifecycle).
pub fn to_decision_record(r: &SessionRecord) -> decision::SessionRecord {
    decision::SessionRecord {
        session_id: r.session_id.clone(),
        spec_path: r.spec_path.clone(),
        spec_summary: r.spec_summary.clone(),
        start_commit: r.start_commit.clone(),
        start_branch: r.start_branch.clone(),
        opened_at: parse_ts(&r.opened_at),
        status: to_decision_status(r.status),
        checkpoints: r.checkpoints.iter().map(to_decision_checkpoint).collect(),
        closed_at: r.closed_at.as_deref().map(parse_ts),
        end_commit: r.end_commit.clone(),
    }
}
